### PR TITLE
add paddle.optimizer.LBFGS API and a modify its test case test=develop

### PR DESCRIPTION
**Pull request by lijialin03**
### PR types

Others
### PR changes

APIs
### Description

Pcard-66961
add paddle.optimizer.LBFGS API
modify numel in _add_grad function to shorten time

### CODE REVIEW

File: python/paddle/incubate/optimizer/lbfgs.py [hunk 1]
```diff
@@ -202,10 +202,16 @@ def _gather_flat_grad(self):
...
         offset = 0
         for p in self._params:
-            numel = p.numel().item()
+            p_shape = p.shape
+            if 0 in p_shape or (-1) in p_shape:
+                numel = p.numel().item()
+            else:
+                numel = 1
+                for dim in p.shape:
+                    numel *= dim
```
Comment (cxxly) [inline python/paddle/incubate/optimizer/lbfgs.py]: functools.reduce(lambda x, y: x*y, x.shape)

File: python/paddle/fluid/tests/unittests/test_lbfgs_v2.py [hunk 2]
```diff
@@ -269,6 +269,33 @@ def func3(x, alpha, d):
...
+                max_eval=None,
+                tolerance_grad=1e-07,
+                tolerance_change=1e-09,
+                history_size=5,
+                line_search_fn='strong_wolfe',
+                parameters=net.parameters(),
+            )
+
+        self.assertRaises(AssertionError, error_func3)
+
```
Comment (cxxly) [inline python/paddle/fluid/tests/unittests/test_lbfgs_v2.py]: move all test case into test_lbfgs.py

**Comment by cxxly** [comment]
add @framework.non_static_only for step

**Comment by cxxly** [comment]
copy same version into python/paddle/optimizer/ and add LFBGS into __all__ of __init__.py

**Comment by cxxly** [comment]
add deprecated for LBFGS in incubate

**Comment by cxxly** [comment]
move all test cast into one file t`test_lbfgs_class.py`

File: python/paddle/fluid/tests/unittests/test_lbfgs_dynamic.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+import numpy as np
+
+import paddle
+from paddle.optimizer import LBFGS
+from paddle.optimizer.lbfgs import _cubic_interpolate, _strong_wolfe
```
Comment (cxxly) [inline python/paddle/fluid/tests/unittests/test_lbfgs_dynamic.py]: import follow pep2 code style, import moudle rather than class/function

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

**Review by cxxly — DISMISSED** [review]
LGTM

**Review by XieYunshen — DISMISSED** [review]
LGTM 
单测重命名

**Review by XieYunshen — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by sunzhongkai588 — APPROVED** [review]
LGTM

**Pull request merged**